pkg/koordlet/runtimehooks/hooks/hooks.go

```diff
@@ -0,0 +1,104 @@
+/*
+ Copyright 2022 The Koordinator Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package hooks
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/klog/v2"
+)
+
+type HookStage string
+
+const (
+	PreRunPodSandbox            HookStage = "PreRunPodSandbox"
+	PreStartContainer           HookStage = "PreStartContainer"
+	PostStartContainer          HookStage = "PostStartContainer"
+	PostStopContainer           HookStage = "PostStopContainer"
+	PreUpdateContainerResources HookStage = "PreUpdateContainerResources"
+)
+
+type Hook struct {
+	name        string
+	stage       HookStage
+	description string
+	fn          HookFn
+}
+
+type HookFn func(request, response interface{}) error
+
+var globalStageHooks map[HookStage][]*Hook
+var globalRWMutex sync.RWMutex
+
+func Register(stage HookStage, name, description string, hookFn HookFn) *Hook {
+	h, error := getNewHook(stage, name)
+	if error != nil {
+		klog.Fatal("hook %s is conflict since name is already registered")
+		return h
+	}
+	h.description = description
+	h.fn = hookFn
+	return h
+}
+
+func getNewHook(stage HookStage, name string) (*Hook, error) {
```

Comment:
this function should be named 'generateNewHook'?

Reply:
updated